File: tutorials/basic/11.latent.variables/model/load_data.py

import numpy as np

class PointsInfoStruct:
    def __init__(self, points, nPoints, nDimensions, nClusters=None, assignments=None ):
        self.points = np.array(points)
        self.nPoints = nPoints
        self.nDimensions = nDimensions
        self.nClusters = nClusters
        self.assignments = np.array(assignments)


    def reset_points(self, new_points, new_assignments=None, new_nClusters=1):
        assert len(new_points) == len(new_assignments)
        self.nPoints = len(new_points)
        self.points = np.array(new_points)
        self.assignments = [] if new_assignments is None else new_assignments
        self.assignments = np.array(self.assignments)
        for pt in new_points:
            assert len(pt) == self.nDimensions
        self.nClusters = new_nClusters



def extr(line, toint=True, tofloat=False):
    result =  line.strip("\n ").split(" ")
    if tofloat:
        result = [float(r) for r in result]
    elif toint:
        result = [int(r) for r in result]
    return result

class UnivariateData(PointsInfoStruct):
   def __init__(self):

       self.assignments = []
       self.points = []
       self.nClusters = 1

       print("Loading data from data_single.in ... \n")
       with open("model/data_single.in", "r") as fd:
           self.nPoints, self.nDimensions = extr(fd.readline())
           for i in range(self.nPoints):
               point = extr(fd.readline(), tofloat=True)
               self.points.append(point)

       self.points = np.array(self.points)


class MultivariateData(PointsInfoStruct):
    def __init__(self):

        self.assignments = []
        self.points = []

        with open("model/data_multiple.in", "r") as fd:
            self.nPoints, self.nDimensions, self.nClusters = extr(fd.readline())
            for i in range(self.nPoints):
                linecontents = extr(fd.readline(), toint=False)
                assert len(linecontents) == self.nDimensions + 1
                self.points.append(list(map(float, linecontents[:-1])))
                self.assignments.append(int(linecontents[-1]))

        self.points = np.array(self.points)
        self.assignments = np.array(self.assignments)

File: features/9.latent.variables.cxx/_model/load_data.cpp

#include "load_data.hpp"
#include "string.h"
#include <limits>
#include <math.h>
#include <stdio.h>
#include <unistd.h>

pointsInfo __p;

pointsInfoStruct &univariateData()
{
  FILE *problemFile;
  FILE *problemFile2;
  unsigned long nPoints = 0;
  unsigned long nDimensions = 0;

  printf("Loading data from data_single.in ... \n");
  problemFile = fopen("_model/data_single.in", "r");
  if (!problemFile)
  {
    char cwd[PATH_MAX];
    getcwd(cwd, sizeof(cwd));
    std::cout << "Current workdir: " << cwd << std::endl;
    fprintf(stderr, "[Error] File not found: model/data_single.in. Make sure you run this script from the 'tutorials/advanced/9.latent.variables' directory. ");
    exit(-1);
  }
  problemFile2 = fopen("data_single.in", "r");
  fscanf(problemFile, "%lu", &nPoints);
  fscanf(problemFile, "%lu", &nDimensions);

  __p.points.resize(nPoints);
  __p.assignments.resize(0);

  for (size_t i = 0; i < nPoints; i++)
  {
    __p.points[i].resize(nDimensions);
    double val;
    for (size_t j = 0; j < nDimensions; j++)
    {
      fscanf(problemFile, "%le ", &val);
      __p.points[i][j] = val;
    }
  }
  fclose(problemFile);
  __p.nDimensions = nDimensions;
  __p.nPoints = nPoints;

  return __p;
};

pointsInfoStruct &multivariateData()
{
  // Also contains assignment information; remove to be sure you are not using it
  FILE *problemFile;

  unsigned long nPoints = 0;
  unsigned long nDimensions = 0;
  unsigned long nClusters = 0;

  printf("Loading data from data_multiple.in ... \n");
  problemFile = fopen("_model/data_multiple.in", "r");
  fscanf(problemFile, "%lu", &nPoints);
  fscanf(problemFile, "%lu", &nDimensions);
  fscanf(problemFile, "%lu", &nClusters);

  __p.points.resize(nPoints);
  __p.assignments.resize(nPoints);
  for (std::vector<double> &p : __p.points)
  {
    p.resize(nDimensions);
  }
  for (size_t i = 0; i < nPoints; i++)
  {
    __p.points[i].resize(nDimensions);
  }

  for (size_t i = 0; i < nPoints; i++)
  {
    double val_d;
    for (size_t j = 0; j < nDimensions; j++)
    {
      fscanf(problemFile, "%le ", &val_d);
      __p.points[i][j] = val_d;
    }
    unsigned long val;
    fscanf(problemFile, "%lu ", &val);
    __p.assignments[i] = val;
  }
  __p.nDimensions = nDimensions;
  __p.nPoints = nPoints;
  __p.nClusters = nClusters;

  fclose(problemFile); // Todo: remove some day. Added this when looking for a very mysterious bug

  return __p;
};

void reset_points(pointsInfoStruct &__p, std::vector<std::vector<double>> new_points, std::vector<int> new_assignments, int new_nClusters)
{
  assert(new_points.size() == new_assignments.size());
  __p.nPoints = new_points.size();
  __p.points = new_points;
  __p.assignments = new_assignments;
  for (std::vector<double> pt : new_points)
    assert(pt.size() == __p.nDimensions && "Cannot change dimensionality (could implement it though)");
  __p.nClusters = new_nClusters;
  return;
}
